File: guests/python/save.py
"""The save round trip (tools/scenes/save.steps). EVERY STATUS IS A READ-BACK
OFF THE DISK, through the HANDLE, and no name carries an extension."""

import os
import pathlib
import sys
import tempfile
import threading

import kaya

app = kaya.App()

# Both halves compute this identically. `tempfile.gettempdir()` and NEVER
# `TMPDIR` — docs/traps.md.
save_dir = pathlib.Path(tempfile.gettempdir()) / f"kaya-save-{os.getpid()}"
save_dir.mkdir(parents=True, exist_ok=True)
# The decoy the picker needs (guests/python/filedialog.py says why):
# "decoy" MUST sort before "draft".
(save_dir / "draft").write_text("first draft")
(save_dir / "decoy").write_text("decoy")

# HANDLES and never paths: the phones have no re-openable path.
source = None
destination = None


def read_back(picked):
    """Read a handle back through kaya, with Python's own file API."""
    try:
        handle, _seekable = picked.open(kaya.wire.FILE_MODE_READ)
    except OSError as e:
        return f"open failed: {e}"
    try:
        with handle as f:
            return f.read().decode()
    except OSError as e:
        return f"read failed: {e}"


def write_back(picked, text):
    """Write `text` through a handle and report what the file says
    afterwards. FILE_MODE_WRITE truncates, on a picked file and on a
    save destination alike — the destination only adds the create."""
    try:
        handle, _seekable = picked.open(kaya.wire.FILE_MODE_WRITE)
    except OSError as e:
        # Without the create a save destination cannot be opened
        # (docs/save-plan.md D1).
        return f"save failed: {e}"
    try:
        # Closed BEFORE the reopen, so the bytes read back are the file's.
        with handle as f:
            f.write(text.encode())
    except OSError as e:
        return f"write failed: {e}"
    return read_back(picked)


def work(job):
    """Run one file operation on a thread of the guest's own and post the
    answer back."""
    def worker():
        text = job()
        app.post(lambda: status.set(text))

    threading.Thread(target=worker, name="save-worker", daemon=True).start()


def picked(files):
    global source
    if not files:
        status.set("open cancelled")
        return
    file = files[0]
    source = file
    work(lambda: f"opened {read_back(file)}")


def saved(file):
    global destination
    if file is None:
        status.set("save cancelled")
        return
    destination = file
    work(lambda: f"saved {write_back(file, 'third draft')}")


def open_file():
    kaya.pick_file(on_result=picked)


def save_back():
    # No dialog: the chosen handle is writable. A missing one gets its OWN
    # sentence, never a crash (docs/deferred.md, save-jvm WATCH).
    file = source
    if file is None:
        status.set("nothing open to save")
        return
    work(lambda: f"saved {write_back(file, 'second draft')}")


def save_as():
    # The name the dialog OPENS with. NO FILTER either: with types set,
    # NSSavePanel appends an extension (docs/deferred.md).
    kaya.save_file("copy", on_result=saved)


def reopen():
    # A save through the wrong handle fails only here.
    first, second = source, destination
    if first is None or second is None:
        status.set("nothing to reopen")
        return
    work(lambda: f"reopened {read_back(first)} {read_back(second)}")


with app.window(title="save"):
    status = kaya.signal("no file")
    with kaya.column():
        kaya.label(bind=status).a11y_id("status")  # label#0
        kaya.button("open", on_click=open_file)  # button#0
        kaya.button("save", on_click=save_back)  # button#1
        kaya.button("save as", on_click=save_as)  # button#2
        kaya.button("reopen", on_click=reopen)  # button#3

sys.exit(app.run())

File: guests/js/save.ts
// The save round trip (tools/scenes/save.steps). EVERY STATUS IS A READ-BACK
// OFF THE DISK, through the HANDLE, and no name carries an extension.

import { mkdirSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import * as kaya from "kaya-gui";

const app = new kaya.App();

// `os.tmpdir()` and NEVER `TMPDIR` — docs/traps.md.
const saveDir = join(tmpdir(), `kaya-save-${process.pid}`);
mkdirSync(saveDir, { recursive: true });
// The decoy MUST sort before "draft" (guests/js/filedialog.ts says why).
writeFileSync(join(saveDir, "draft"), "first draft", { encoding: "utf-8" });
writeFileSync(join(saveDir, "decoy"), "decoy", { encoding: "utf-8" });

// HANDLES and never paths: the phones have no re-openable path.
let source: kaya.PickedFile | null = null;
let destination: kaya.PickedFile | null = null;

/** What a failed file operation says about itself. */
function reason(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}

/** Read a handle back through kaya, with the guest's own file API. */
function readBack(picked: kaya.PickedFile): string {
  // The addon reads over the platform handle (docs/js-plan.md §6).
  try {
    return new TextDecoder().decode(picked.read());
  } catch (e) {
    return `read failed: ${reason(e)}`;
  }
}

function writeBack(picked: kaya.PickedFile, text: string): string {
  // What the dialog handed back opens EMPTY (docs/save-plan.md D1).
  try {
    picked.write(text);
  } catch (e) {
    return `write failed: ${reason(e)}`;
  }
  return readBack(picked);
}

/** One file operation, with the ANSWER posted: this worker has no second
 * thread to hand the job to. */
function work(job: () => string): void {
  const text = job();
  app.post(() => status.set(text));
}

function picked(files: kaya.PickedFile[]): void {
  if (files.length === 0) {
    status.set("open cancelled");
    return;
  }
  const file = files[0]!;
  source = file;
  work(() => `opened ${readBack(file)}`);
}

function saved(file: kaya.PickedFile | null): void {
  if (file === null) {
    status.set("save cancelled");
    return;
  }
  destination = file;
  work(() => `saved ${writeBack(file, "third draft")}`);
}

async function openFile(): Promise<void> {
  picked(await kaya.pickFile());
}

function saveBack(): void {
  // No dialog: the chosen handle is writable. A missing one gets its OWN
  // sentence, never a crash (docs/deferred.md, save-jvm WATCH).
  const file = source;
  if (file === null) {
    status.set("nothing open to save");
    return;
  }
  work(() => `saved ${writeBack(file, "second draft")}`);
}

async function saveAs(): Promise<void> {
  // The name the dialog OPENS with. NO FILTER either: with types set,
  // NSSavePanel appends an extension (docs/deferred.md).
  saved(await kaya.saveFile("copy"));
}

function reopen(): void {
  // A save through the wrong handle fails only here.
  const first = source;
  const second = destination;
  if (first === null || second === null) {
    status.set("nothing to reopen");
    return;
  }
  work(() => `reopened ${readBack(first)} ${readBack(second)}`);
}

let status!: kaya.Signal<string>;

app.window({ title: "save" }, () => {
  status = kaya.signal("no file");
  kaya.column(() => {
    kaya.label({ bind: status }).a11yId("status"); // label#0
    kaya.button("open", { onClick: openFile }); // button#0
    kaya.button("save", { onClick: saveBack }); // button#1
    kaya.button("save as", { onClick: saveAs }); // button#2
    kaya.button("reopen", { onClick: reopen }); // button#3
  });
});

app.run();
